pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -0,0 +1,366 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.AllowInsecure, err = p.configVarResolver.GetConfigVarBoolValue(rawConfig.AllowInsecure)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.CPUs = rawConfig.CPUs
+	c.MemoryMB = rawConfig.MemoryMB
+
+	return &c, &pconfig, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	config, _, err := p.getConfig(spec.ProviderConfig)
+	client, err := getClient(config.Username, config.Password, config.VSphereURL, config.AllowInsecure)
+	if err != nil {
+		return fmt.Errorf("failed to get vsphere client: '%v'", err)
+	}
+
+	finder, err := getDatacenterFinder(config.Datacenter, client)
+	if err != nil {
+		return err
+	}
+
+	_, err = finder.Datastore(context.TODO(), config.Datastore)
+	if err != nil {
+		return err
+	}
+
+	_, err = finder.ClusterComputeResource(context.TODO(), config.Cluster)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, userdata string) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	client, err := getClient(config.Username, config.Password, config.VSphereURL, config.AllowInsecure)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get vsphere client: '%v'", err)
+	}
+
+	vmID, err := CreateLinkClonedVm(machine.Spec.Name,
+		config.TemplateVMName,
+		config.Datacenter,
+		config.Cluster,
+		config.CPUs,
+		config.MemoryMB,
+		client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create linked vm: '%v'", err)
+	}
+
+	finder, err := getDatacenterFinder(config.Datacenter, client)
+	if err != nil {
+		return nil, err
+	}
+	virtualMachine, err := finder.VirtualMachine(context.TODO(), machine.Spec.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get virtual machine object: %v", err)
+	}
+
+	localUserdataIsoFilePath, err := generateLocalUserdataIso(userdata, machine.Spec.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := os.Remove(localUserdataIsoFilePath)
+		if err != nil {
+			glog.Errorf("failed to clean up local userdata iso file at %s: %v", localUserdataIsoFilePath, err)
+		}
+	}()
+
+	err = uploadAndAttachISO(finder, virtualMachine, localUserdataIsoFilePath, config.Datastore, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to upload and attach userdata iso: %v", err)
+	}
+
+	// Ubuntu wont boot with attached floppy device, because it tries to write to it
+	// which fails, because the floppy device does not contain a floppy disk
+	// Upstream issue: https://bugs.launchpad.net/cloud-images/+bug/1573095
+	err = removeFloppyDevice(virtualMachine)
+	if err != nil {
+		return nil, err
+	}
+
+	powerOnTask, err := virtualMachine.PowerOn(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("failed to power on machine: %v", err)
+	}
+	powerOnTask.Wait(context.TODO())
```

Comment:
should we set some sane timeout for this operation ?

Reply:
I guess we could, but it would require some effort since that is not supported natively, so we would have to write some wrapper that calls `powerOnTask.Wait` in a dedicated goroutine.

Do you think that's worth it? Or is there an easier approach?

Reply:
we could use a context that has timeout set.
https://golang.org/pkg/context/#WithTimeout